### PR TITLE
AoT-integration of new blockchain hook api

### PR DESCRIPTION
**Pull request by sasurobert**
integration of new blockchain hook api

### CODE REVIEW

**Review by andrei-marinica — DISMISSED** [review]

File: arwen/contexts/runtime.go [hunk 1]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
+	if context.useWarmInstance {
+		context.warmInstanceAddress = context.GetSCAddress()
+		context.warmInstance = context.instance
+	}
+
+	if newCode || len(codeHash) == 0 {
+		codeHash, err = context.host.Crypto().Sha256(contract)
 		if err != nil {
-			context.instance = nil
+			context.CleanWasmerInstance()
```
Comment (camilbancioiu) [inline arwen/contexts/runtime.go:174]: Need to set `warmInstance` and `warmInstanceAddress` to `nil` as well (assuming the instance was properly set, but somehow hashing fails?).

File: arwen/contexts/runtime.go [hunk 2]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
 
-		idContext := arwen.AddHostContext(context.host)
-		context.instance.SetContextData(idContext)
+	idContext := arwen.AddHostContext(context.host)
+	context.instance.SetContextData(idContext)
 
+	if newCode {
 		err = context.VerifyContractCode()
 		if err != nil {
 			context.CleanWasmerInstance()
```
Comment (camilbancioiu) [inline arwen/contexts/runtime.go:187]: Also need to set `warmInstance` and `warmInstanceAddress` to `nil` here.

File: arwen/contexts/runtime.go [hunk 3]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
+	if warmInstanceSet {
+		return nil
+	}
+
+	compiledCodeSet := context.makeInstanceFromCompiledCode(codeHash, gasLimit, newCode)
+	if compiledCodeSet {
+		return nil
+	}
+
+	log.Trace("Creating a new Wasmer instance")
```
Comment (camilbancioiu) [inline arwen/contexts/runtime.go:149]: Optional: the rest of this function, starting here, could be put into a separate private method `makeInstanceFromContractBytecode()`.

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

File: arwen/contexts/runtime.go [hunk 4]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
-		gasSchedule := context.host.Metering().GasSchedule()
-		options := wasmer.CompilationOptions{
-			GasLimit:           gasLimit,
-			UnmeteredLocals:    uint64(gasSchedule.WASMOpcodeCost.LocalsUnmetered),
-			OpcodeTrace:        false,
-			Metering:           true,
-			RuntimeBreakpoints: true,
-		}
-		newInstance, err := wasmer.NewInstanceWithOptions(contract, options)
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:85]: warm instance is intended to be used only in scquery. will add the new func call.

File: arwen/contexts/runtime.go (hunk 1 above)
Comment (sasurobert) [reply arwen/contexts/runtime.go:174]: warmInstance is for scQuery only - there is no way for hashing to fail. the current implementation does not return error at all.

File: arwen/contexts/runtime.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:85]: done

File: arwen/contexts/runtime.go [hunk 5]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
+func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uint64, newCode bool) error {
+	if context.RunningInstancesCount() >= context.maxWasmerInstances {
+		context.instance = nil
+		return arwen.ErrMaxInstancesReached
+	}
+
+	blockchain := context.host.Blockchain()
+	codeHash := blockchain.GetCodeHash(context.GetSCAddress())
+
+	warmInstanceSet := context.setWarmInstanceWhenNeeded()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:139]: done

File: arwen/contexts/runtime.go [hunk 6]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
+	if warmInstanceSet {
+		return nil
+	}
+
+	compiledCodeSet := context.makeInstanceFromCompiledCode(codeHash, gasLimit, newCode)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:144]: done

File: arwen/contexts/runtime.go [hunk 7]
```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
...
+}
+
+func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uint64, newCode bool) error {
+	if context.RunningInstancesCount() >= context.maxWasmerInstances {
+		context.instance = nil
+		return arwen.ErrMaxInstancesReached
+	}
+
+	blockchain := context.host.Blockchain()
+	codeHash := blockchain.GetCodeHash(context.GetSCAddress())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:137]: done

File: arwen/contexts/runtime.go (hunk 1 above)
Comment (sasurobert) [reply arwen/contexts/runtime.go:174]: moved to the end of function the wasm instance creation.

File: arwen/contexts/runtime.go (hunk 2 above)
Comment (sasurobert) [reply arwen/contexts/runtime.go:187]: done

File: arwen/contexts/runtime.go (hunk 3 above)
Comment (sasurobert) [reply arwen/contexts/runtime.go:149]: done

**Review by andrei-marinica — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**